From 59b856fba451fb2dab043e04c8b43ecefd1c5985 Mon Sep 17 00:00:00 2001
From: Jamie Iles <jamie@jamieiles.com>
Date: Thu, 21 Apr 2011 11:50:23 +0100
Subject: [PATCH 103/228] macb: quirk for undetectable TSU/TSU clk

Some platforms without design configuration registers have no way to
signal that the platform has a TSU and optionally a TSU clk.  Provide
quirks to allow the platform to force this.

Signed-off-by: Jamie Iles <jamie@jamieiles.com>
---
 drivers/net/macb.c                 |    6 ++++--
 include/linux/platform_data/macb.h |    2 ++
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/net/macb.c b/drivers/net/macb.c
index ca60490..ea4f225 100644
--- a/drivers/net/macb.c
+++ b/drivers/net/macb.c
@@ -1005,10 +1005,12 @@ static int macb_configure_tsu(struct macb *bp)
 	int err;
 
 	/* MACB doesn't have a TSU, but GEM does. */
-	if (!bp->is_gem || !(gem_readl(bp, DCFG5) & GEM_BIT(TSU)))
+	if (!bp->is_gem || (!(gem_readl(bp, DCFG5) & GEM_BIT(TSU)) &&
+			    !(bp->quirks & MACB_QUIRK_HAVE_TSU)))
 		return 0;
 
-	if (gem_readl(bp, DCFG5) & GEM_BIT(TSUCLK))
+	if ((gem_readl(bp, DCFG5) & GEM_BIT(TSUCLK)) ||
+	    (bp->quirks & MACB_QUIRK_HAVE_TSU_CLK))
 		bp->tsu = clk_get(&bp->pdev->dev, "tsu");
 	else
 		bp->tsu = bp->pclk;
diff --git a/include/linux/platform_data/macb.h b/include/linux/platform_data/macb.h
index dd4037f..d273feb 100644
--- a/include/linux/platform_data/macb.h
+++ b/include/linux/platform_data/macb.h
@@ -11,6 +11,8 @@
 enum macb_quirks {
 	MACB_QUIRK_NO_UNALIGNED_TX = (1 << 1),
 	MACB_QUIRK_FORCE_DBW64 = (1 << 2),
+	MACB_QUIRK_HAVE_TSU = (1 << 3),
+	MACB_QUIRK_HAVE_TSU_CLK = (1 << 4),
 };
 
 struct macb_platform_data {
-- 
1.7.0.4

